optaplanner-examples/src/test/java/org/optaplanner/examples/nqueens/solver/score/NQueensConstraintProviderTest.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.nqueens.solver.score;
+
+import org.junit.Test;
+import org.optaplanner.core.api.score.buildin.simple.SimpleScore;
+import org.optaplanner.examples.nqueens.domain.NQueens;
+import org.optaplanner.examples.nqueens.domain.Queen;
+import org.optaplanner.examples.nqueens.domain.Row;
+import org.optaplanner.test.impl.score.stream.ConstraintProviderVerifier;
+
+import static org.optaplanner.test.impl.score.stream.ConstraintVerifier.forConstraint;
+import static org.optaplanner.test.impl.score.stream.ConstraintVerifier.forConstraintProvider;
+
+public class NQueensConstraintProviderTest {
+
+    // Dummy data, the tests don't actually do anything.
+    private Queen queen1, queen2;
+    private Row row1, row2;
+
+    @Test
+    public void testSingleConstraint() {
+        NQueensConstraintProvider constraintProvider = new NQueensConstraintProvider();
+        forConstraint(constraintProvider::ascendingDiagonalConflict)
+                .givenFacts(queen1, queen2, row1, row2)
+                .expectImpact(SimpleScore.ONE)  // No type check here. Wrong score type fails at runtime.
```

Comment:
All ScoreVerifier checks should be regardless of the constraingWeight. Only the matchWeight needs to be tested.
This is for both fixed constraintWeight and penalizeConfigurable constraintWeights etc.

Reply:
Actually, I disagree.

We should assert on the final impact, not on the `matchWeight`. You want to know what becomes your score, not some partial calculation that is `matchWeight`.

In fact, that is why I added the method `givenPlanningSolution`, where the configurable constraint weights are defined - so that you can get the final result, not just the match weight.

Reply:
I tend to agree with Geoffrey. In the context of a constraint provider unit test, I'd like to focus on testing purely the constraint behavior, that is:
- whether it's a penalty of a reward,
- how many times it triggers (typically to verify number of matching unique pairs in the given set of objects),
- whether the [`matchWeigher`](https://github.com/kiegroup/optaplanner/blob/7.34.0.Final/optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java#L768) calculates the expected match weight (in cases it is affected by matching object properties, e.g. distance).

I would argue that `@ConstraintWeight` affects total score but not constraint behavior. So, if constraint weight configuration changes, constraints' behavior doesn't change, therefore the constraint provider unit test shouldn't fail.

In other words, I'd like to freely change my constraint weight configuration without breaking CP tests.